x/pubsub/keeper/msg_server.go

```diff
@@ -0,0 +1,360 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+	"github.com/peggyjv/sommelier/v4/x/pubsub/types"
+)
+
+var _ types.MsgServer = Keeper{}
+
+func (k Keeper) AddPublisherIntent(c context.Context, msg *types.MsgAddPublisherIntent) (*types.MsgAddPublisherIntentResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	publisherIntent := msg.PublisherIntent
+
+	if err := publisherIntent.ValidateBasic(); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrInvalid, "invalid publisher intent: %s", err.Error())
+	}
+
+	publisher, found := k.GetPublisher(ctx, publisherIntent.PublisherDomain)
+	if !found {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrNotFound, "no publisher found with domain: %s", publisherIntent.PublisherDomain)
+	}
+
+	signer := msg.MustGetSigner()
+	if signer.String() != publisher.Address {
+		return nil, sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "registered publisher address must be signer: %s", publisher.Address)
+	}
+
+	_, found = k.GetPublisherIntent(ctx, publisherIntent.SubscriptionId, publisherIntent.PublisherDomain)
+	if found {
+		// TODO(bolten): would this UX be better if it removed the old one for you? deleting a publisher intent will clear out
+		// any subscriber intents attached to it
+		return nil, sdkerrors.Wrapf(types.ErrAlreadyExists, "publisher already has intent for this subscription ID, must be removed first")
+	}
+
+	k.SetPublisherIntent(ctx, *publisherIntent)
+
+	ctx.EventManager().EmitEvents(
```

Comment:
I know we never do this anywhere, but EmitEvents and EmitEvent are deprecated and we should be using EmitTypedEvent and EmitTypedEvents which define the events in protos

Reply:
I believe typed events are in newer versions of the SDK than we are still on.